Automatically create a link when selected text is a URL

noisysocks opened the pull request:
Closes #3513. 

When a URL is selected and the Link button is clicked or the Link keyboard shortcut is pressed, a link with that URL set as its `href` should be instantly created.

![auto-link](https://user-images.githubusercontent.com/612155/44248249-c5777080-a22c-11e8-9834-b7facdc8502e.gif)

As part of this, I pulled an existing link regular expression out into its own `isURL` function that lives in the `@wordpress/url` package. Not sure if it's actually a useful helper function, though—happy to go back on this decision.

**To test:**

```
npm run test-e2e
```

Or, for those of us who distrust automation:

1. Highlight a URL
2. Click _Link_ or press <kbd>Cmd</kbd>+<kbd>K</kbd>
3. The selected URL should immediately become a link

noisysocks committed:
Automatically create a link when selected text is a URL

When a URL is selected and the Link button is clicked or the Link
keyboard shortcut is pressed, a link with that URL set as its href
should be instantly created.

jorgefilipecosta commented on packages/url/src/index.js:
```diff
@@ -3,9 +3,21 @@
+ *
+ * @return {boolean} Whether or not it looks like a URL.
+ */
+export function isURL( url ) {
```
Nice util, I think we may be able to use it in:
https://github.com/WordPress/gutenberg/blob/f9da76a18f1dda3b0aaa00218b49914901c6ab0e/packages/editor/src/components/rich-text/index.js#L285
Removing a duplicate regex.

noisysocks replied:
I've already replaced that regex 🙂

jorgefilipecosta reviewed (commented):
Nice feature 👍 
It is working well I think we have a small bug that we should address before the merge. After we add a link using this method if we click right away on the edit button nothing happens the edit button does not work. If we select other block and then come back and press edit things work well.

noisysocks committed:
Simplify isURL tests

noisysocks committed:
Fix URL link editing

Fixes a bug that prevented the user from selecting a link which has text
that looks like a URL and clicking Edit. We accomplish this by
distinguishing between editing a link (isEditingLink) and adding a new
link (isAddingLink).

noisysocks commented on packages/url/src/test/index.test.js:
```diff
@@ -1,7 +1,42 @@
+		];
+
+		forEach( urls, ( url ) => {
+			expect( isURL( url ) ).toBe( true );
```
👍 0d218ad96

noisysocks commented:
Good catch @jorgefilipecosta! I've fixed this in 50c9c01b0.

jorgefilipecosta reviewed (approved):
It is working well in my tests 👍

mtias commented:
Thanks for this improvement!
